ops/framework.py

```diff
@@ -88,7 +88,48 @@ def from_path(cls, path):
         return handle
 
 
-class EventBase:
+class UnassignedEvent(type):
+    """Metaclass for events which have not yet been assigned to an emitter type.
+
+    Without being assigned and bound to an emitter, unassigned events cannot be emitted.
+    """
+
+    def __set_name__(event_type, emitter_type, event_kind):
+        # Permanently assign the event for this emitter type.
```

Comment:
This comment is not very precise either. What the logic below does is to replace the event type by an event descriptor that knows about its context. Again, no need to document it. It's one line and it's very readable.